monitoring/metric/counter.go

```diff
@@ -0,0 +1,99 @@
+// Copyright 2017 Google Inc. All Rights Reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package metric
+
+import (
+	"context"
+	"sort"
+	"sync"
+	"time"
+
+	"github.com/golang/glog"
+)
+
+type Counter interface {
+	Add(n int64)
+}
+type counter struct {
+	mu              sync.Mutex
+	name            string
+	value           int64
+	lastDumped      time.Time
+	lastDumpedValue int64
+}
+
+type safeMetrics struct {
+	mu sync.Mutex
+	m  map[string]*counter
+}
+
+var (
+	metrics = safeMetrics{
+		m: make(map[string]*counter),
+	}
+)
+
+func (m *counter) Add(n int64) {
+	m.mu.Lock()
+	defer m.mu.Unlock()
+	m.value += n
+}
+
+func NewCounter(name string) Counter {
+	c := counter{name: name, lastDumped: time.Now()}
+	metrics.mu.Lock()
+	defer metrics.mu.Unlock()
+	if dup := metrics.m[c.name]; dup != nil {
+		glog.Fatal("duplicate metric name registered: ", c.name)
+	}
+	metrics.m[c.name] = &c
+	return &c
+}
+
+func dump() {
+	metrics.mu.Lock()
+	defer metrics.mu.Unlock()
+	glog.Info("dumping metrics:")
+	keys := make([]string, 0, len(metrics.m))
+	for k, _ := range metrics.m {
+		keys = append(keys, k)
+	}
+	sort.Strings(keys)
+	for _, key := range keys {
+		m := metrics.m[key]
+		m.mu.Lock()
+		defer m.mu.Unlock()
+		delta := m.value - m.lastDumpedValue
+		now := time.Now()
+		duration := now.Sub(m.lastDumped)
+		qps := float64(delta) / duration.Seconds()
```

Comment:
I see this as part of the "dumping to log backend", rather than the metric, but they're slightly intertwined for now. In the case of a Prometheus handler, we wouldn't do that at all, we'd just give out the current value, and let Prometheus calculate rates.